### PR TITLE
Update Amcharts auto-update config and meta data

### LINKED ISSUE #5337: [Request] update amcharts
Amcharts
http://www.amcharts.com/javascript-charts/changelog/

3.15.2 is out

### PR DESCRIPTION
**Pull request by maruilian11**
**git repo url:** https://github.com/amcharts/amcharts3
**Watch 4
  Star 7
 Fork 3**
@Amomo  could u help me check this PR for #5337 ? thank you!
- [ ] Not found on cdnjs repo
- [x] No already exist issue and PR
- [ ] Notable popularity
- [x] Project has public repository on famous online hosting platform (or been hosted on npm) 
- [x] Has valid tags for each versions (for git auto-update)
- [x] Auto-update setup
- [x] Auto-update target/source is valid.
- [x] Auto-update filemap is correct

### CODE REVIEW

**Comment by Amomo** [comment]
Then, I think you can combine `"add author field"` and `"add repository"` to one commit.
Also, `"update the keywords"` and `"update license"`.

**Comment by maruilian11** [comment]
@Amomo could you help me check the PR again? thank you!

**Comment by Amomo** [comment]
@maruilian11 Everything is fine, but 5943664.
As you have add auto-update config in package.json now.
I think we don't need to add a exist lib manually.
Could you remove this commit? Thank you!

**Comment by Amomo** [comment]
And, could you `cc` the author account in the commit message?
It would be good to notify him this lib is adding to cdnjs!

**Comment by maruilian11** [comment]
@Amomo please cc #5753 you will know why i add with a new version, besides, i forgot to add requiredfiles before, thanks for reminding.

**Comment by Amomo** [comment]
ok, I see the problem. Thank you for your mention.

**Comment by maruilian11** [comment]
@Amomo could you help me check this PR again? thank you!

**Comment by Amomo** [comment]
@PeterDaveHello I think this PR is fine, could you take a look?
Thank you!

**Comment by PeterDaveHello** [comment]
What's the reason why to add real files here? Should leave them alone.

**Comment by maruilian11** [comment]
@PeterDaveHello cc#5753 Amomo has the same question above.

**Comment by PeterDaveHello** [comment]
I only see
> @PeterDaveHello I think this PR is fine, could you take a look? Thank you!

**Comment by PeterDaveHello** [comment]
#5753 did show your "same question"

**Comment by maruilian11** [comment]
@PeterDaveHello does the same question means i still shouldn't add with the new version?

**Comment by PeterDaveHello** [comment]
You can add it.

**Comment by Amomo** [comment]
Hi @maruilian11 I am so sorry to keep you waiting so long.:cold_sweat:
Could you re-send the commits until this PR pass the test?

Thank you very much!

**Comment by maruilian11** [comment]
@Amomo finally~ it pass :joy:

**Comment by Amomo** [comment]
@maruilian11 Thank you!

@PeterDaveHello I think this PR is fine.
In the first commit, maruilian11 add the real files through the discussion in #5753.
Could you take a look? Thank you! :-D

**Comment by PeterDaveHello** [comment]
Can you talk about the steps to add the files in this PR? And I think we need a double review in the cases that adding/touching many files. @LeaYeh please join.

**Comment by maruilian11** [comment]
sure. first i made pre-version [3.13.0](https://github.com/cdnjs/cdnjs/tree/compatibleTest/ajax/libs/amcharts/3.13.0) on cdnjs as example to help choose which files could be added.
the difference between 3.13.0 and 3.17.1 is following:
1.there is no `exporting` folder in v3.17.1
2.there are two new folder (`dataloader` and `export`) in `plugins`, so i add them by the same way as [responsive](https://github.com/cdnjs/cdnjs/tree/compatibleTest/ajax/libs/amcharts/3.13.0/plugins/responsive).
3.there is a new `.js` (gantt.js) in v3.17.1.

ps. about the reason i add real file, cc #5753 
is that clear? 
@PeterDaveHello @Amomo @LeaYeh

**Comment by Amomo** [comment]
I think the explanation is okay, but... amchart is already v3.17.3 now. 😨
@maruilian11 could you update your PR? Thank you very much!

**Comment by maruilian11** [comment]
done.

**Comment by maruilian11** [comment]
@Amomo thank you for your reminding!

**Comment by Amomo** [comment]
@PeterDaveHello I think this PR is fine, could you take a look? Thank you!

**Comment by maruilian11** [comment]
@LeaYeh could you help me do the double check? thank you~

**Comment by maruilian11** [comment]
@LeaYeh ：)

**Comment by LeaYeh** [comment]
checked.
@PeterDaveHello I think this PR is fine, could you take a look?

**Comment by PeterDaveHello** [comment]
The `basePath` - `amcharts` looks like its dist folder, maybe we could use a simply auto-update on it, current config is a little bit too complex and not be so reliable. You guys would you please confirm that? Thanks.

**Comment by maruilian11** [comment]
@PeterDaveHello but even we regard the folder as dist, because of the `pulgins` contains files we don't need, we still can't use `**/*.+(js|css|map)` for its filemap to make it less complex.

**Comment by PeterDaveHello** [comment]
You meant which file? I think the current rules are too complex, maybe we should ask for a disk folder or a npm package, maybe the author will help.

**Comment by maruilian11** [comment]
@PeterDaveHello like `amcharts/plugins/export/examples/export.config.advanced.js` https://github.com/amcharts/amcharts3/blob/master/amcharts/plugins/export/examples/export.config.advanced.js

the author said they do have dist folder on npm package, but its struture is same as `amchart` on github.

**Comment by PeterDaveHello** [comment]
Then switch to npm auto-update and use `**/*`.

**Comment by maruilian11** [comment]
@PeterDaveHello but i can't use `**/*.+(map|css|js)`
could you suggest me some easier auto-update configs?

**Comment by PeterDaveHello** [comment]
ping @cdnjs/intern

**Comment by maruilian11** [comment]
@cdnjs/intern

**Comment by maruilian11** [comment]
@PeterDaveHello seems no response?

**Comment by maruilian11** [comment]
ping @cdnjs/intern

**Comment by maruilian11** [comment]
@cdnjs/intern any advice?

**Comment by PeterDaveHello** [comment]
@maruilian11 since we dropped `"requiredFiles"`, let's simply update `package.json` and add auto-update config here. The problem about you can't use `**/*.+(js|css|map)`, read [document  of **node-glob**](https://github.com/isaacs/node-glob#glob-primer), it's not so hard to exclude some patterns.

File: ajax/libs/amcharts/package.json [hunk 1]
```diff
@@ -1,14 +1,46 @@
...
+      "plugins/responsive/responsive*.+(js|map)",
+      "themes/*.js",
+      "amcharts*.js",
+      "funnel*.js",
+      "gantt*.js",
+      "gauge*.js",
+      "pie*.js",
+      "radar*.js",
+      "serial*.js",
+      "xy*.js"
```
Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: @PeterDaveHello can i  use 
`"(images|lang|patterns|themes)+(/**/*.)+(png|gif|svg|jpg|js)",`
`"plugins/*/*.+(js|css|map)",`
`"*.js"`
as file map?

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: Can't we use something like `**/*!(blah)*` here?

Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: how about `**/*!(examples/**/*|libs/**/*|license.txt|LICENSE|readme.md|README.md)*
i'm not sure if these folders under plugins are needed:
https://github.com/amcharts/amcharts3/tree/master/amcharts/plugins/export/shapes
https://github.com/amcharts/amcharts3/tree/master/amcharts/plugins/export/lang

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: Soooooo ... complex, please try to make it be easier to understand, you don't need to force combine them into one rule.

Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: how about 
`+(image|lang|patterns|themes)/**/*.+(js|gif|png|svg|jpg)`
`plugins/**/!(LICENSE|examples/**/*|*.txt|*.md)`
`*.js`

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: Still too complex in the second line.

Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: @PeterDaveHello maybe you can give me some advice.

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: you don't need to force combine them into one rule

Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: you mean this line?
`plugins/**/!(LICENSE|examples/**/*|libs/**/*|*.txt|*.md)`

i think combine them would be convenient for me.

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: Current config is hard to read ...

Comment (maruilian11) [inline ajax/libs/amcharts/package.json]: yes, maybe it is, so can you tell what config will you set directly?

Comment (PeterDaveHello) [inline ajax/libs/amcharts/package.json]: I'll take a look at it later, you can solve other issues first.

**Comment by PeterDaveHello** [comment]
Let's finish this ASAP.

1. Separate the rules, seems the rules could be complex if we try to combine them.
2. Drop the assets in the commit, I'll handle it to make sure images would been pre-compressed.
3. Rebase this branch with the latest master branch.


@cdnjs/intern2 please also help review this huge PR, thanks.

Thanks.

**Comment by maruilian11** [comment]
@cdnjs/intern2 please you help me review this PR, thanks.

**Comment by pvnr0082t** [comment]
@PeterDaveHello @cdnjs/intern @cdnjs/intern2 I got some errors as following after running git auto-update with config of amcharts:
```
All files for amcharts v3.20.17 - 196
196 true
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern1r.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern1r.jpg' }
Some strange error occured here
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern2.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern2.jpg' }
Some strange error occured here
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern3.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern3.jpg' }
Some strange error occured here
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern4.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern4.jpg' }
Some strange error occured here
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern5.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern5.jpg' }
Some strange error occured here
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern6.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern6.jpg' }
```
Though I got these errors, it can still work. I'm not sure if it's my problem. Will these errors affected anything?

**Comment by PeterDaveHello** [comment]
@pvnr0082t would you please confirm if the files been mentioned still there as we need?

**Comment by pvnr0082t** [comment]
@PeterDaveHello Yes, the files we need are still there.

**Comment by PeterDaveHello** [comment]
@pvnr0082t I mean, the same same file? should confirm via sha1

**Comment by pvnr0082t** [comment]
@PeterDaveHello Yes, I confirmed using `sha1sum` and they are the same.

File: ajax/libs/amcharts/package.json [hunk 2]
```diff
@@ -2,13 +2,34 @@
...
+  "autoupdate": {
+    "source": "git",
+    "target": "git://github.com/amcharts/amcharts3.git",
+    "fileMap": [
+      {
+        "basePath": "amcharts",
+        "files": [
+          "patterns/**/*",
+          "**/!(examples)/!(LICENSE|index.js|*.json|*.md|*.txt|*.csv|*.png)",
+          "*.+(js|map)"
```
Comment (x09326) [inline ajax/libs/amcharts/package.json]: I think we can just use `line 29` and remove `line 30`, `line 29` will get the files of `line 30` right?

Comment (pvnr0082t) [inline ajax/libs/amcharts/package.json]: No. Have you tried it?

**Comment by maruilian11** [comment]
@Piicksarn @cdnjs/intern2 I rebased the branch with the latest master branch.
Do I need to do any change? Or any advice? Thanks!

**Comment by pvnr0082t** [comment]
LGTM

**Comment by PeterDaveHello** [comment]
@kennynaoh please help review this PR, thanks.

**Comment by pvnr0082t** [comment]
ping @kennynaoh

**Comment by kennynaoh** [comment]
@maruilian11 Sorry for late response. Please rebase to the latest master branch. Thanks!

**Comment by PeterDaveHello** [comment]
@kennynaoh please tell about your reviewer result before asking for rebase.

**Comment by kennynaoh** [comment]
I think this PR is ok.

**Comment by maruilian11** [comment]
@Piicksarn @cdnjs/intern2 I rebase the branch with the latest master branch. Could you help me review this PR? Thanks!

**Comment by pvnr0082t** [comment]
LGTM

**Comment by PeterDaveHello** [comment]
Why do we need to exclude the png images? 

https://github.com/amcharts/amcharts3/tree/3.20.9/amcharts/patterns/black

**Comment by maruilian11** [comment]
@PeterDaveHello 
if we use `**/!(examples)/!(LICENSE|index.js|*.json|*.md|*.txt|*.csv)`, `png`s under `plugins/export/examples/images` won't be ignore.
https://github.com/amcharts/amcharts3/tree/master/amcharts/plugins/export/examples/images

how about use `**/!(examples|patterns)/!(LICENSE|index.js|*.json|*.md|*.txt|*.csv|*.png)` to not exclude the `png`s under `patterns/*`

**Comment by PeterDaveHello** [comment]
I didn't get it, do we need it or not? Can you confirm with its author?

**Comment by maruilian11** [comment]
@PeterDaveHello I think we do need it.
you said
>Why do we need to exclude the png images?
>
>https://github.com/amcharts/amcharts3/tree/3.20.9/amcharts/patterns/black

Actually, we do include those png images by using this filemap:
`"patterns/**/*"`

The second rule won't exclude the png images under `patterns/black`:
`"patterns/**/*",`
` "**/!(examples)/!(LICENSE|index.js|*.json|*.md|*.txt|*.csv|*.png)",`

I think we both think we need those png images, so I didn't confirm it with the author.
Do you still need me to check it?

**Comment by PeterDaveHello** [comment]
So the reason you think we need it just by my asking? Thanks funny!

**Comment by maruilian11** [comment]
@cdnjs/intern @cdnjs/intern2 I have confirmed the rules in [issue](https://github.com/amcharts/amcharts3/issues/92) opened in amcharts.
they recommend we avoid adding `samples` folder as well, so I add it to the rules.
Please help me review this PR.

@PeterDaveHello amcharts members hope we can also add these two repos:
https://github.com/amcharts/amstock3
https://github.com/amcharts/ammap3

But those repos popularity isn't enough, do we also add them?

**Comment by PeterDaveHello** [comment]
Ping @x09326 @kennynaoh

**Comment by x09326** [comment]
LGTM

**Comment by PeterDaveHello** [comment]
@pvnr0082t is the problem you mentioned solved?

```
{ [Error: ENOENT: no such file or directory, unlink '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern1r.jpg']
  errno: -2,
  code: 'ENOENT',
  syscall: 'unlink',
  path: '/home/p/cdnjs/ajax/libs/amcharts/3.20.17/patterns/chalk/pattern1r.jpg' }
```

@maruilian11 @kennynaoh @x09326 don't you guys have that problem? Or you don't even run the auto-update? I didn't see any further discussion about this part.

I think we got a big problem on "discussion", please treat it seriously, sometimes you just ignore the problem been mentioned and then just say something like "I have fixed it", "LGTM", that makes no sense, we should discuss and make the best decision that we can make.

**Comment by pvnr0082t** [comment]
@PeterDaveHello Yes, that's the problem I met.

**Comment by kennynaoh** [comment]
Yes, I have the problem, too.

**Comment by x09326** [comment]
I meet the same problem, too. But I don't know what's the problem, so I don't know how to discuss it then I just wait for the answer

**Comment by PeterDaveHello** [comment]
@kennynaoh @x09326 if you don't know what's going on and you have the same problem:

1. At least report that you have the same problem
2. Help valid if we still get correct files by sha256 and md5sum, as I asked @pvnr0082t 

Need to think more about what have you tried, and I think all of us will need this kind of discussion and help from the others someday.

@maruilian11 don't you have the problem?

**Comment by maruilian11** [comment]
@PeterDaveHello @pvnr0082t mentioned the problem before https://github.com/cdnjs/cdnjs/pull/5747#issuecomment-261776100
And we checked the files mentioned in auto-update still can be got.

I think the problem is 
`"**/!(examples|samples)/!(LICENSE|index.js|*.json|*.md|*.txt|*.csv|*.png)"` can only catch `patterns/chalk/*`, but it can't catch `patterns/white/*` and `patterns/black/*` because there is only `png`s under them.
and
`patterns/**` will catch `chalk` `white` and `black`.
so `chalk` has been double got. I think that's why the error message comes.

there is a way to avoid error message and get the files we want:
use `patterns/!(chalk)/**` instead of `patterns/**`
I checked the files we want are still there.

sorry for not changing the rule before because I thought we talk this problem before already.

File: ajax/libs/amcharts/package.json [hunk 3]
```diff
@@ -2,13 +2,34 @@
...
+    "url": "https://github.com/amcharts/amcharts3.git"
+  },
+  "autoupdate": {
+    "source": "git",
+    "target": "git://github.com/amcharts/amcharts3.git",
+    "fileMap": [
+      {
+        "basePath": "amcharts",
+        "files": [
+          "patterns/!(chalk)/**",
```
Comment (pvnr0082t) [inline ajax/libs/amcharts/package.json]: @maruilian11 I think we can just use single star like `"patterns/!(chalk)/*",`. thank you.

**Comment by pvnr0082t** [comment]
Okay, the error not happened after avoid double fetching files under `patterns/chalk`.

**Comment by maruilian11** [comment]
@cdnjs/intern2 I updated the auto-update filemap. Please help me review this PR, thanks!

**Comment by PeterDaveHello** [comment]
ping @kennynaoh

**Comment by kennynaoh** [comment]
Sorry for late response. I think this PR is ok. Please rebase to the latest master branch. Thanks!

**Comment by maruilian11** [comment]
@cdnjs/intern2 I just rebased this branch with our latest master, thanks!

**Comment by kennynaoh** [comment]
LGTM

**Comment by PeterDaveHello** [comment]
@maruilian11 @pvnr0082t @kennynaoh please be careful about the PR title, I thought you guys always forget this part.

**Pull request merged**